src/steppable.cr

```diff
@@ -0,0 +1,157 @@
+# Implements a `#step` method for iterating from a value.
+module Steppable
+  # Iterates from `self` to *limit* incrementing by the amount of *step* on each
+  # iteration.
+  # If *exclusive* is `true`, *limit* is excluded from the iteration.
+  #
+  # ```
+  # ary = [] of Int32
+  # 1.step(to: 4, by: 2) do |x|
+  #   ary << x
+  # end
+  # ary                                        # => [1, 3]
+  # 1.step(to: 4, by: 2).to_a                  # => [1, 3]
+  # 1.step(to: 4, by: 1).to_a                  # => [1, 2, 3, 4]
+  # 1.step(to: 4, by: 1, exclusive: true).to_a # => [1, 2, 3]
+  # ```
+  #
+  # The type of each iterated element is `typeof(self + step)`.
+  #
+  # If *to* is `nil`, iteration is open ended.
+  #
+  # The starting point (`self`) is always iterated as first element, with two
+  # exceptions:
+  # * if `self` and *to* don't compare (i.e. `(self <=> to).nil?`). Example:
+  #   `1.0.step(Float::NAN)`
+  # * if the direction of *to* differs from the direction of `by`. Example:
+  #   `1.step(to: 2, by: -1)`
+  #
+  # In those cases the iteration is empty.
+  def step(*, to limit = nil, by step, exclusive : Bool = false, &) : Nil
+    # type of current should be the result of adding `step`:
+    current = self + (step - step)
+
+    if limit == current
+      # Only yield current if it's also the limit.
+      # Step size doesn't matter in this case: `1.step(to: 1, by: 0)` yields `1`
+      yield current unless exclusive
+      return
+    end
+
+    raise ArgumentError.new("Zero step size") if step.zero?
```

Comment:
Maybe mention that this method relies on `zero?` and `sign` to work?

Reply:
It also relies on meaningful implementations of `<=>`, `+` and `-`. Not sure it's worth listing all of them, because describing that would probably not be more concise than looking at the code.